File: deepeval/metrics/contextual_relevancy/contextual_relevancy.py
from typing import Optional, List, Type, Union
import asyncio

from deepeval.utils import get_or_create_event_loop, prettify_list
from deepeval.metrics.utils import (
    construct_verbose_logs,
    trimAndLoadJson,
    check_llm_test_case_params,
    initialize_model,
)
from deepeval.test_case import (
    LLMTestCase,
    LLMTestCaseParams,
    ConversationalTestCase,
)
from deepeval.metrics import BaseMetric
from deepeval.models import DeepEvalBaseLLM
from deepeval.metrics.contextual_relevancy.template import (
    ContextualRelevancyTemplate,
)
from deepeval.metrics.indicator import metric_progress_indicator
from deepeval.metrics.contextual_relevancy.schema import *


class ContextualRelevancyMetric(BaseMetric):
    _required_params: List[LLMTestCaseParams] = [
        LLMTestCaseParams.INPUT,
        LLMTestCaseParams.ACTUAL_OUTPUT,
        LLMTestCaseParams.RETRIEVAL_CONTEXT,
    ]

    def __init__(
        self,
        threshold: float = 0.5,
        model: Optional[Union[str, DeepEvalBaseLLM]] = None,
        include_reason: bool = True,
        async_mode: bool = True,
        strict_mode: bool = False,
        verbose_mode: bool = False,
        evaluation_template: Type[
            ContextualRelevancyTemplate
        ] = ContextualRelevancyTemplate,
    ):
        self.threshold = 1 if strict_mode else threshold
        self.model, self.using_native_model = initialize_model(model)
        self.evaluation_model = self.model.get_model_name()
        self.include_reason = include_reason
        self.async_mode = async_mode
        self.strict_mode = strict_mode
        self.verbose_mode = verbose_mode
        self.evaluation_template = evaluation_template

    def measure(
        self,
        test_case: LLMTestCase,
        _show_indicator: bool = True,
        _in_component: bool = False,
    ) -> float:

        check_llm_test_case_params(test_case, self._required_params, self)

        self.evaluation_cost = 0 if self.using_native_model else None
        with metric_progress_indicator(
            self, _show_indicator=_show_indicator, _in_component=_in_component
        ):
            if self.async_mode:
                loop = get_or_create_event_loop()
                loop.run_until_complete(
                    self.a_measure(
                        test_case,
                        _show_indicator=False,
                        _in_component=_in_component,
                    )
                )
            else:
                self.verdicts_list: List[ContextualRelevancyVerdicts] = [
                    (self._generate_verdicts(test_case.input, context))
                    for context in test_case.retrieval_context
                ]
                self.score = self._calculate_score()
                self.reason = self._generate_reason(test_case.input)
                self.success = self.score >= self.threshold
                self.verbose_logs = construct_verbose_logs(
                    self,
                    steps=[
                        f"Verdicts:\n{prettify_list(self.verdicts_list)}",
                        f"Score: {self.score}\nReason: {self.reason}",
                    ],
                )

            return self.score

    async def a_measure(
        self,
        test_case: LLMTestCase,
        _show_indicator: bool = True,
        _in_component: bool = False,
    ) -> float:

        check_llm_test_case_params(test_case, self._required_params, self)

        self.evaluation_cost = 0 if self.using_native_model else None
        with metric_progress_indicator(
            self,
            async_mode=True,
            _show_indicator=_show_indicator,
            _in_component=_in_component,
        ):
            self.verdicts_list: List[ContextualRelevancyVerdicts] = (
                await asyncio.gather(
                    *[
                        self._a_generate_verdicts(test_case.input, context)
                        for context in test_case.retrieval_context
                    ]
                )
            )
            self.score = self._calculate_score()
            self.reason = await self._a_generate_reason(test_case.input)
            self.success = self.score >= self.threshold
            self.verbose_logs = construct_verbose_logs(
                self,
                steps=[
                    f"Verdicts:\n{prettify_list(self.verdicts_list)}",
                    f"Score: {self.score}\nReason: {self.reason}",
                ],
            )

            return self.score

    async def _a_generate_reason(self, input: str):
        if self.include_reason is False:
            return None

        irrelevant_statements = []
        relevant_statements = []
        for verdicts in self.verdicts_list:
            for verdict in verdicts.verdicts:
                if verdict.verdict.lower() == "no":
                    irrelevant_statements.append(verdict.reason)
                else:
                    relevant_statements.append(verdict.statement)

        prompt: dict = self.evaluation_template.generate_reason(
            input=input,
            irrelevant_statements=irrelevant_statements,
            relevant_statements=relevant_statements,
            score=format(self.score, ".2f"),
        )
        if self.using_native_model:
            res, cost = await self.model.a_generate(prompt, schema=Reason)
            self.evaluation_cost += cost
            return res.reason
        else:
            try:
                res: Reason = await self.model.a_generate(prompt, schema=Reason)
                return res.reason
            except TypeError:
                res = await self.model.a_generate(prompt)
                data = trimAndLoadJson(res, self)
                return data["reason"]

    def _generate_reason(self, input: str):
        if self.include_reason is False:
            return None

        irrelevant_statements = []
        relevant_statements = []
        for verdicts in self.verdicts_list:
            for verdict in verdicts.verdicts:
                if verdict.verdict.lower() == "no":
                    irrelevant_statements.append(verdict.reason)
                else:
                    relevant_statements.append(verdict.statement)

        prompt: dict = self.evaluation_template.generate_reason(
            input=input,
            irrelevant_statements=irrelevant_statements,
            relevant_statements=relevant_statements,
            score=format(self.score, ".2f"),
        )
        if self.using_native_model:
            res, cost = self.model.generate(prompt, schema=Reason)
            self.evaluation_cost += cost
            return res.reason
        else:
            try:
                res: Reason = self.model.generate(prompt, schema=Reason)
                return res.reason
            except TypeError:
                res = self.model.generate(prompt)
                data = trimAndLoadJson(res, self)
                return data["reason"]

    def _calculate_score(self):
        total_verdicts = 0
        relevant_statements = 0
        for verdicts in self.verdicts_list:
            for verdict in verdicts.verdicts:
                total_verdicts += 1
                if verdict.verdict.lower() == "yes":
                    relevant_statements += 1

        if total_verdicts == 0:
            return 0

        score = relevant_statements / total_verdicts
        return 0 if self.strict_mode and score < self.threshold else score

    async def _a_generate_verdicts(
        self, input: str, context: List[str]
    ) -> ContextualRelevancyVerdicts:
        prompt = self.evaluation_template.generate_verdicts(
            input=input, context=context
        )
        if self.using_native_model:
            res, cost = await self.model.a_generate(
                prompt, schema=ContextualRelevancyVerdicts
            )
            self.evaluation_cost += cost
            return res
        else:
            try:
                res = await self.model.a_generate(
                    prompt, schema=ContextualRelevancyVerdicts
                )
                return res
            except TypeError:
                res = await self.model.a_generate(prompt)
                data = trimAndLoadJson(res, self)
                return ContextualRelevancyVerdicts(**data)

    def _generate_verdicts(
        self, input: str, context: str
    ) -> ContextualRelevancyVerdicts:
        prompt = self.evaluation_template.generate_verdicts(
            input=input, context=context
        )
        if self.using_native_model:
            res, cost = self.model.generate(
                prompt, schema=ContextualRelevancyVerdicts
            )
            self.evaluation_cost += cost
            return res
        else:
            try:
                res = self.model.generate(
                    prompt, schema=ContextualRelevancyVerdicts
                )
                return res
            except TypeError:
                res = self.model.generate(prompt)
                data = trimAndLoadJson(res, self)
                return ContextualRelevancyVerdicts(**data)

    def is_successful(self) -> bool:
        if self.error is not None:
            self.success = False
        else:
            try:
                self.success = self.score >= self.threshold
            except:
                self.success = False
        return self.success

    @property
    def __name__(self):
        return "Contextual Relevancy"

File: typescript/src/metrics/contextual-relevancy/contextual-relevancy.ts
import { BaseMetric, resolveThreshold } from "@/metrics/base-metrics";
import { LLMTestCase, SingleTurnParams } from "@/test-case";
import { DeepEvalBaseLLM } from "@/models";
import type { DeepEvalBaseSystemOneModel } from "@/models/system-one";
import type { EvalModeName } from "@/config/eval-mode";
import {
  initializeMetricModels,
  generateWithSchema,
  constructVerboseLogs,
  prettifyList,
  resolveRetrievalContext,
} from "@/metrics/utils";
import {
  generateQagVerdicts,
  parseQuestions,
  runSystemOneEval,
  splitSentences,
  type SystemOneEvalSpec,
  type SystemOneVerdictSpec,
} from "@/metrics/system-one";
import {
  ContextualRelevancyVerdictsSchema,
  ContextualRelevancyScoreReasonSchema,
  type ContextualRelevancyVerdict,
  type ContextualRelevancyVerdicts,
} from "@/metrics/contextual-relevancy/schema";
import { type MetricTemplateOverride } from "@/templates/override";
import { contextualRelevancyVerdictVars } from "@/metrics/retrieval-context-display";
import { prepareMeasure } from "@/metrics/prepare-measure";

const TEMPLATE_CLASS = "ContextualRelevancyMetric";

export type ContextualRelevancyTemplateOverride =
  MetricTemplateOverride<"ContextualRelevancyMetric">;

function systemOneRelevancyVerdict(
  statement: string,
  verdict: string,
  probability: number,
): ContextualRelevancyVerdict {
  let reason = `P(yes)=${probability.toFixed(2)}`;
  if (verdict === "no") {
    reason = `'${statement}' is not relevant to the input (${reason}).`;
  }
  return { statement, verdict, reason };
}

export interface ContextualRelevancyMetricOptions {
  threshold?: number | null;
  flaky?: boolean;
  model?: DeepEvalBaseLLM | string;
  /** The System One model (Jev) used under `hybrid` / `system_one`. */
  systemOneModel?: DeepEvalBaseSystemOneModel | string;
  /** Who decides; defaults to `DEEPEVAL_EVAL_MODE`, then `llm`. */
  evalMode?: EvalModeName;
  includeReason?: boolean;
  strictMode?: boolean;
  verboseMode?: boolean;
  showIndicator?: boolean;
  evaluationTemplate?: ContextualRelevancyTemplateOverride;
}

/**
 * Contextual Relevancy — what fraction of statements across `retrievalContext`
 * are relevant to the `input`? Judge each node's statements, then
 * score = relevant statements / total statements.
 */
export class ContextualRelevancyMetric extends BaseMetric {
  verdictsList: ContextualRelevancyVerdicts[] = [];

  constructor(options: ContextualRelevancyMetricOptions = {}) {
    const strictMode = options.strictMode ?? false;
    super(strictMode ? 1 : resolveThreshold(options.threshold, 0.5), {
      strictMode,
      verboseMode: options.verboseMode,
      includeReason: options.includeReason ?? true,
      showIndicator: options.showIndicator,
      flaky: options.flaky,
      evaluationTemplate: options.evaluationTemplate,
    });
    this.multimodalAware = true;
    this.templateClass = TEMPLATE_CLASS;
    this.requiredParams = [
      SingleTurnParams.INPUT,
      SingleTurnParams.RETRIEVAL_CONTEXT,
    ];
    initializeMetricModels(this, options);
  }

  async measure(testCase: LLMTestCase): Promise<number> {
    this.error = undefined;
    await this.startProgress();
    try {
      prepareMeasure(this, testCase);
      if (await runSystemOneEval(this, testCase)) return this.score as number;

      const retrievalContext = resolveRetrievalContext(
        testCase.retrievalContext ?? [],
      );
      this.verdictsList = await Promise.all(
        retrievalContext.map((context) =>
          this.generateVerdicts(testCase.input, context),
        ),
      );
      this.score = this.calculateScore();
      this.reason = await this.generateReason(testCase.input);
      this.success = this.isSuccessful();

      this.verboseLogs = constructVerboseLogs(this, [
        `Verdicts:\n${prettifyList(this.verdictsList)}`,
        `Score: ${this.score}\nReason: ${this.reason}`,
      ]);
      return this.score;
    } finally {
      this.stopProgress();
    }
  }

  systemOneEvalSpec(testCase: LLMTestCase): SystemOneEvalSpec | undefined {
    if (testCase.multimodal) return undefined;
    return {
      evaluationParams: this.requiredParams,
      questions: parseQuestions(
        this.getPrompt("_experimental_system_one_questions"),
      ),
    };
  }

  private systemOneVerdictSpec(
    input: string,
    context: string,
  ): SystemOneVerdictSpec<string, ContextualRelevancyVerdict> | undefined {
    if (this.multimodal) return undefined;
    return {
      instructions: this.getPrompt("_experimental_system_one_verdict"),
      items: splitSentences(context),
      itemKey: "statement",
      state: { input },
      buildVerdict: systemOneRelevancyVerdict,
    };
  }

  private async generateVerdicts(
    input: string,
    context: string,
  ): Promise<ContextualRelevancyVerdicts> {
    const verdicts = await generateQagVerdicts(this, {
      systemOne: this.systemOneVerdictSpec(input, context),
      llm: async () => {
        const prompt = this.getPrompt("generate_verdicts", {
          input,
          context,
          ...contextualRelevancyVerdictVars(this.multimodal),
        });
        const { verdicts } = await generateWithSchema(
          this,
          prompt,
          ContextualRelevancyVerdictsSchema,
        );
        return verdicts;
      },
    });
    return { verdicts };
  }

  private async generateReason(input: string): Promise<string | undefined> {
    if (!this.includeReason) return undefined;
    const irrelevantStatements: (string | null | undefined)[] = [];
    const relevantStatements: string[] = [];
    for (const { verdicts } of this.verdictsList) {
      for (const v of verdicts) {
        if (v.verdict.toLowerCase() === "no")
          irrelevantStatements.push(v.reason);
        else relevantStatements.push(v.statement);
      }
    }
    const prompt = this.getPrompt("generate_reason", {
      input,
      irrelevant_statements: irrelevantStatements,
      relevant_statements: relevantStatements,
      score: (this.score ?? 0).toFixed(2),
    });
    const { reason } = await generateWithSchema(
      this,
      prompt,
      ContextualRelevancyScoreReasonSchema,
    );
    return reason;
  }

  private calculateScore(): number {
    let totalVerdicts = 0;
    let relevant = 0;
    for (const { verdicts } of this.verdictsList) {
      for (const v of verdicts) {
        totalVerdicts++;
        if (v.verdict.toLowerCase() === "yes") relevant++;
      }
    }
    if (totalVerdicts === 0) return 0;
    const score = relevant / totalVerdicts;
    return this.applyStrictMode(score);
  }

  get name(): string {
    return "Contextual Relevancy";
  }
}
